src/storage/src/source/generator/key_value.rs

```diff
@@ -0,0 +1,584 @@
+                Ok(SourceMessage {
+                    key: Row::pack_slice(&[Datum::UInt64(key)]),
+                    value: Row::pack_slice(&[Datum::UInt64(partition), Datum::Bytes(value_buffer)]),
+                    metadata: if self.include_offset {
+                        Row::pack(&[Datum::UInt64(self.next_offset)])
+                    } else {
+                        Row::default()
+                    },
+                }),
+            );
+            (msg, MzOffset::from(self.next_offset), 1)
+        }));
+
+        // Advance to the next offset.
+        self.next_offset += 1;
+        self.next_offset
+    }
+}
+
+pub fn render_statistics_operator<G: Scope<Timestamp = MzOffset>>(
+    scope: &mut G,
+    config: RawSourceCreationConfig,
+    resume_uppers: impl futures::Stream<Item = Antichain<MzOffset>> + 'static,
+) -> (Stream<G, ProgressStatisticsUpdate>, PressOnDropButton) {
+    let id = config.id;
+    let mut builder =
+        AsyncOperatorBuilder::new(format!("key_value_loadgen_statistics:{id}"), scope.clone());
+
+    let (mut stats_output, stats_stream) = builder.new_output();
+
+    let button = builder.build(move |caps| async move {
+        let [stats_cap]: [_; 1] = caps.try_into().unwrap();
+
+        let offset_worker = config.responsible_for(0);
+
+        if !offset_worker {
+            // Emit 0, to mark this worker as having started up correctly.
+            stats_output
+                .give(
+                    &stats_cap,
+                    ProgressStatisticsUpdate::SteadyState {
+                        offset_known: 0,
+                        offset_committed: 0,
+                    },
+                )
+                .await;
+            return;
+        }
+
+        tokio::pin!(resume_uppers);
+        loop {
+            match resume_uppers.next().await {
+                Some(frontier) => {
+                    if let Some(offset) = frontier.as_option() {
+                        stats_output
+                            .give(
+                                &stats_cap,
+                                ProgressStatisticsUpdate::SteadyState {
+                                    offset_known: offset.offset,
+                                    offset_committed: offset.offset,
+                                },
+                            )
+                            .await;
+                    }
+                }
+                None => return,
+            }
+        }
+    });
+
+    (stats_stream, button.press_on_drop())
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[mz_ore::test]
+    fn test_key_value_loadgen_resume_upper() {
+        let up = UpdateProducer::new(
+            1,     // partition 1
+            3,     // of 3 partitions
+            126,   // 126 keys.
+            2,     // 2 snapshot rounds
+            0,     // not important
+            2,     // batch size 2
+            1234,  //seed
+            5,     // resume upper of 5
```

Comment:
Yeah.. we should definitely be passing the struct with the named fields here

Reply:
yeah I agree!